packages/vite/src/plugins/plugin.ts

```diff
@@ -23,76 +28,123 @@ export interface VitePluginOptions {
   previewTargetName?: string;
   serveStaticTargetName?: string;
 }
+type ViteTargets = Pick<ProjectConfiguration, 'targets' | 'metadata'>;
 
 const cachePath = join(workspaceDataDirectory, 'vite.hash');
 const targetsCache = readTargetsCache();
 
-function readTargetsCache(): Record<
-  string,
-  Record<string, TargetConfiguration>
-> {
-  return existsSync(cachePath) ? readJsonFile(cachePath) : {};
+function readTargetsCache(
+  cache: string = cachePath
+): Record<string, ViteTargets> {
+  return existsSync(cache) ? readJsonFile(cache) : {};
 }
 
-function writeTargetsToCache() {
-  const oldCache = readTargetsCache();
-  writeJsonFile(cachePath, {
-    ...oldCache,
-    ...targetsCache,
-  });
+function writeTargetsToCache(
+  cache: string = cachePath,
+  results?: Record<string, ViteTargets>
+) {
+  if (!results) {
+    writeJsonFile(cache, results);
+  } else {
+    const oldCache = readTargetsCache();
+    writeJsonFile(cachePath, {
+      ...oldCache,
+      ...targetsCache,
+    });
+  }
 }
 
 export const createDependencies: CreateDependencies = () => {
   writeTargetsToCache();
   return [];
 };
 
+const viteVitestConfigGlob = '**/{vite,vitest}.config.{js,ts,mjs,mts,cjs,cts}';
+
+export const createNodesV2: CreateNodesV2<VitePluginOptions> = [
+  viteVitestConfigGlob,
+  async (configFilePaths, options, context) => {
+    const optionsHash = hashObject(options);
+    const cachePath = join(workspaceDataDirectory, `vite-${optionsHash}.hash`);
+    const targetsCache = readTargetsCache(cachePath);
+    try {
+      return await createNodesFromFiles(
+        (configFile, options, context) =>
+          createNodesInternal(configFile, options, context, targetsCache),
+        configFilePaths,
+        options,
+        context
+      );
+    } finally {
+      writeTargetsToCache(cachePath, targetsCache);
+    }
+  },
+];
+
 export const createNodes: CreateNodes<VitePluginOptions> = [
-  '**/{vite,vitest}.config.{js,ts,mjs,mts,cjs,cts}',
+  viteVitestConfigGlob,
   async (configFilePath, options, context) => {
-    const projectRoot = dirname(configFilePath);
-    // Do not create a project if package.json and project.json isn't there.
-    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
-    if (
-      !siblingFiles.includes('package.json') &&
-      !siblingFiles.includes('project.json')
-    ) {
-      return {};
-    }
+    logger.warn(
+      '`createNodes` is deprecated. Update your plugin to utilize createNodesV2 instead. In Nx 20, this will change to the createNodesV2 API.'
+    );
+    return createNodesInternal(configFilePath, options, context, targetsCache);
```

Comment:
It's okay to use an `{}` here for `targetsCache`.

This function shouldn't really get called unless it's a rare case that someone wrapped our plugin `createNodes` function.